### PR TITLE
Django: Capture custom request/response headers

### LINKED ISSUE #913: Django: Capture request/response headers as span attributes
part of #906

### PR DESCRIPTION
**Pull request by sanketmehta28**
# Description

This PR https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md for django framework

Fixes #913

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?
Added unit tests separately for request and response headers for ASGI and WSGI applications. Also tested it manually by capturing the request and response headers from a instrumented django app for both ASGI and WSGI application

Same test cases for both ASGI/WSGI application
- [x] Test A: check for request header which is set in OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_SERVER_REQUEST env variable, is present in the span as attribute
- [x] Test B: check for request header which is NOT set in OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_SERVER_REQUEST env variable, is NOT present in the span as attribute
- [x] Test C: check for response header which is set in OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_SERVER_RESPONSE env variable, is present in the span as attribute
- [x] Test D: check for response header which is NOT set in OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_SERVER_RESPONSE env variable, is NOT present in the span as attribute

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

**Comment by srikanthccv** [comment]
Please update the title

**Comment by sanketmehta28** [comment]
> Please update the title

done

**Comment by sanketmehta28** [comment]
Hi @open-telemetry/python-approvers : please do review this PR

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,9 +25,6 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [1.10.0-0.29b0](https://github.com/open-telemetry/opentelemetry-python/releases/tag/v1.10.0-0.29b0) - 2022-03-10
 
-- `opentelemetry-instrumentation-wsgi` Capture custom request/response headers in span attributes
```
Comment (lzchen) [inline CHANGELOG.md:25]: Did this not make the previous release?

Comment (sanketmehta28) [reply CHANGELOG.md:25]: yes but the position of this statement was wrong so I just corrected it

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 2]
```diff
@@ -271,12 +291,27 @@ def process_response(self, request, response):
         if activation and span:
             if is_asgi_request:
                 set_status_code(span, response.status_code)
+                if span.is_recording() and span.kind == SpanKind.SERVER:
+                    # asgi_getter inside asgi_collect_custom_response_attributes
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:295]: I don't believe you have to explain this reasoning behind this, so I would remove the comments.

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py [hunk 3]
```diff
@@ -271,12 +291,27 @@ def process_response(self, request, response):
...
+                    # requires 'headers' key to fetch the actuals headers where
+                    # Django response object has 'headers' properties.
+                    # Thats why I have to create a separate dict with 'headers'
+                    # key to make it compatible with asgi_getter
+
+                    custom_headers = {"headers": response.items()}
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:301]: If the key needs to be matched with `AsgiGetter`, we should probably use a constant variable.

Comment (sanketmehta28) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:301]: You mean in place of custom_headers or "headers"?

Comment (sanketmehta28) [reply instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:296]: The thing is for asgi framework, "headers" works fine as the the ASGI request and response object has "scope" dict which has the key "headers" by default.
the problem occurs only for django as its request and response object has headers as properties not as dict. so I have used a temporary variable " custom_header" to make the asgi_getter function work consistently.

**Review by lzchen — APPROVED** [review]

**Pull request merged**